### DISCUSSION TITLE
[RFC] New Component: Float Button

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#37406] [RFCs - archive]
<a name="Summary"></a>
## Summary
添加新组件 FloatButton，默认固定于页面右下角，类似于 antd 官网上的主题按钮：<br />
![image](https://user-images.githubusercontent.com/27722486/188411261-2a6adae5-3385-463a-89d6-2e453b16128d.png)

<a name="Motivation"></a>
## Motivation
当下许多网站都会选择提供一些明显的功能，希望用户可以在浏览网站时无论在什么位置都可以注意到，往往就会是一个浮动按钮的方式放在右下角。<br />在 Ant Design 中我们有一个相似的组件 BackTop，其功能是当用户滚动页面后提供一键返回顶部的能力。它的定位和 FloatButton 相似，但是功能是固定的。所以我们希望提供一个更加通用的组件，把 BackTop 收敛进这个组件中。

<a name="API"></a>
## API
<a name="McdJM"></a>
### FloatButton
```tsx
type FloatButtonProps = {
  className?: string;
  // 可以支持用 top/bottom 等定位修改位置
  style?: React.CSSProperties;
  // 图标
  icon?: React.ReactNode;
  // 文字及其他内容
  description?: React.ReactNode;
  // 变体
  type?: 'default' | 'primary';
  // 形状
  shape?: 'circle' | 'square';
  // Tooltip 内容
  tooltip?: React.ReactNode;
  // 事件
  onClick?: React.MouseEvent<HTMLButtonElement>;
  // 链接
  href?: string;
  target?: string;
}
```
<a name="dvGqM"></a>
### FloatButton.Group
```tsx
type FloatButtonGroupProps = FloatButtonProps {
  // 形状
  shape?: 'circle' | 'square';
  // 包含的 Float Button
  children: React.ReactElement;
  // 触发方式 (有触发方式为菜单模式）
  trigger?: 'click' | 'hover';
  // 受控展开
  open?: boolean;
  // 展开收起的回调
  onOpenChange?: (open: boolean) => void;
}
```
<a name="Sk42K"></a>
### FloatButton.BackTop
```tsx
type FloatButtonBackTopProps = FloatButtonProps & BackTopProps;
```
<a name="9afe3fae"></a>
## Basic Example
<a name="de09e0ca"></a>
### 基础样式
![image](https://user-images.githubusercontent.com/27722486/188441621-a0d32101-257d-45a1-8d59-c19fe2f59d23.png)
```tsx
import { FloatButton } from 'antd';
import { FileTextOutlined } from '@ant-design/icons';

export default () => (
  <>
    <FloatButton icon={<FileTextOutlined />} />
    <FloatButton icon={<FileTextOutlined />} type="primary" />
    <FloatButton.BackTop />
  </>
)
```

<a name="Pa2po"></a>
### 方形按钮
![image](https://user-images.githubusercontent.com/27722486/188441652-b5c91f29-3936-4553-8df6-46325ff0da34.png)
```tsx
import { FloatButton } from 'antd';
import { FileTextOutlined } from '@ant-design/icons';

export default () => (
  <>
    <FloatButton icon={<FileTextOutlined />} shape="square" description="文档" />
    <FloatButton shape="square" description="帮助文档" />
    <FloatButton icon={<FileTextOutlined />} shape="square" description="帮助文档" />
  </>
)
```

<a name="Qz57E"></a>
### 含有气泡卡片的 FloatButton
![image](https://user-images.githubusercontent.com/27722486/188441669-5df8e485-ad12-4eea-b843-737da8b47d88.png)
```tsx
import { FloatButton } from 'antd';
import { FileTextOutlined } from '@ant-design/icons';

export default () => (
  <FloatButton icon={<FileTextOutlined />} tooltip="帮助文档" />
)
```

<a name="ZBrt2"></a>
### 浮动按钮组
![image](https://user-images.githubusercontent.com/27722486/188411346-1cc5373d-7398-4568-9d3a-98fde8b3494f.png)
![image](https://user-images.githubusercontent.com/27722486/188411366-550e3ff7-fa59-4e67-91c3-49faa676cb68.png)

```tsx
import { FloatButton } from 'antd';
import { FileTextOutlined, QuestionCircleOutlined, CodeOutlined } from '@ant-design/icons';

export default () => (
  <>
    <FloatButton.Group>
      <FloatButton icon={<QuestionCircleOutlined />} />
      <FloatButton icon={<FileTextOutlined />} tooltip="帮助文档" />
      <FloatButton.BackTop />
    </FloatButton.Group>
    <FloatButton.Group shape="square">
      <FloatButton icon={<QuestionCircleOutlined />} />
      <FloatButton icon={<FileTextOutlined />} tooltip="帮助文档" />
      <FloatButton icon={<CodeOutlined />} />
      <FloatButton.BackTop />
    </FloatButton.Group>
  </>
)
```

<a name="gR44F"></a>
### 菜单模式
![image](https://user-images.githubusercontent.com/27722486/188441701-24d3e19d-38d1-4dc4-bd37-890ffe91bf96.png)
```tsx
import { FloatButton } from 'antd';
import { FileTextOutlined, CustomerServiceOutlined, CommentOutlined } from '@ant-design/icons';

export default () => (
  <FloatButton.Group 
    icon={<CustomerServiceOutlined />}
    type="primary"
    trigger="click"
  />
    <FloatButton icon={<FileTextOutlined />} tooltip="帮助文档" />
    <FloatButton icon={<CommentOutlined />} />
  </FloatButton.Group>
)
```
<a name="g8f1U"></a>
## 
<a name="d24309fc"></a>
## Detailed Design

- 原来的 BackTop 组件的属性仍然支持，并且新的 FloatButton 的属性也可以作用于 BackTop。
- 菜单模式下触发方式为 `hover` 时，热区应适当放大，防止鼠标移出 FloatButton 后触发关闭了，从而点击不到菜单。
- 自动避让的功能无法实现，让用户使用 `style`自定义 FloatButton 的位置即可。

<a name="sW4mg"></a>
### 设计图
![image](https://user-images.githubusercontent.com/27722486/188441716-888faa6a-bf2a-481f-949b-6f6a2032c196.png)

<a name="Drawbacks"></a>
## Drawbacks

No obvious drawbacks yet.

**Comment by li-jia-nan** [comment]
mark~

**Comment by afc163** [comment]
设计稿有点模糊。。

**Comment by afc163** [comment]
<img width="727" alt="图片" src="https://user-images.githubusercontent.com/507615/189286712-af181b14-c84c-4c81-96fa-35b6b5a8f66f.png">

这个图我和下面的代码对不起来。有上面这个图里的实现案例么

**Reply by MadCcc** [reply]
这个是一个点击展开的 case，group 实体化了

**Reply by afc163** [reply]
那就是 Group 可以嵌套 Group？

**Reply by MadCcc** [reply]
不推荐，这个菜单已经可以看作一个 Group 了，如果塞到其他 Group 里定位肯定有问题

**Comment by afc163** [comment]
```
<FloatButton icon={<FileTextOutlined />} shape="square" description="文档" />
```

这个如果要配置成一个链接，如何写？

**Reply by MadCcc** [reply]
加 `href`，和 Button 一样，我补充一下

**Reply by afc163** [reply]
那还有 target

**Comment by afc163** [comment]
<img width="154" alt="图片" src="https://user-images.githubusercontent.com/507615/189305893-edee175b-5db8-4e87-b660-59985db74e26.png">

这种设计上和实现上如何实现？

**Comment by li-jia-nan** [comment]
可以贴个设计稿的链接嘛？尺寸、圆角、阴影、间距啥的看看具体 value

**Reply by MadCcc** [reply]
![Float Button](https://user-images.githubusercontent.com/27722486/189331849-2699f409-b4a6-41a6-99bc-ea9f1ef0e0a3.png)

**Reply by MadCcc** [reply]
这个好像清楚点，下面有标记间距
基础圆角用 `token.radiusBase` 就行，记得看看 cssinjs 开发指南
https://github.com/ant-design/ant-design/wiki/CSS-in-JS-%E8%BF%81%E7%A7%BB%E6%8C%87%E5%8D%97

**Reply by li-jia-nan** [reply]
明白了

**Comment by li-jia-nan** [comment]
- #37520

**Comment by Veveue** [comment]
@MadCcc  可以支持拖动否
